### PR TITLE
Add UTF-8 friendly `strchr`

### PR DESCRIPTION
**Pull request by Rot127**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Adds a UTF-8 friendly `strchr` and some additional functions which check for UTF-8 character encodings.

This PR is a dependency for https://github.com/rizinorg/rizin/pull/2806 and https://github.com/rizinorg/rizin/pull/2805.

If https://github.com/rizinorg/rizin/pull/2805 is merged. This one can be closed.

**Test plan**

Added.

**Closing issues**

None

### CODE REVIEW

**Review by XVilka — COMMENTED** [review]
These functions could easily cause read OOB though.

File: librz/util/mem.c [hunk 1]
```diff
@@ -20,7 +20,7 @@ RZ_API int rz_mem_count(const ut8 **addr) {
 	return i;
 }
 
-RZ_API int rz_mem_eq(ut8 *a, ut8 *b, int len) {
+RZ_API int rz_mem_eq(const ut8 *a, const ut8 *b, int len) {
```
Comment (wargio) [inline librz/util/mem.c:31]: Please add doxygen

**Comment by Rot127** [comment]
> These functions could easily cause read OOB though.

Added a check for the possible `rz_mem_cmp` OOB reads. The `isXutf8` functions shouldn't have OOB reads, since the if statements don't execute the expressions if the previous fails. Or is this implementation specific?

**Review by XVilka — APPROVED** [review]

File: librz/util/str.c [hunk 2]
```diff
@@ -693,6 +694,96 @@ RZ_API const char *rz_sub_str_rchr(const char *str, int start, int end, char chr
...
+}
+
+/**
+ * \brief Checks if the given character string is a four byte UTF-8 character.
+ *
+ * \param c The character string to test.
+ * \return bool True if the character string is a four byte UTF-8 character. False otherwise.
+ */
+RZ_API bool rz_str_is4utf8(RZ_NULLABLE const char *c) {
+	if (!c[0] || !c[1] || !c[2] || !c[3]) {
```
Comment (wargio) [inline librz/util/str.c]: if c is nullable then when null this will crash

Comment (XVilka) [reply librz/util/str.c]: Yes, I suggest to mark them as `RZ_NONNULL` and add an assert

**Review by wargio — CHANGES_REQUESTED** [review]
fix all null dereferences

**Review by XVilka — APPROVED** [review]

**Review by ret2libc — APPROVED** [review]
LGTM

**Review by wargio — APPROVED** [review]

**Pull request merged**